package/MDAnalysis/lib/mdamath.py

```diff
@@ -158,96 +160,241 @@ def _angle(a, b):
+        else:
+            cos_beta = np.cos(np.deg2rad(beta))
+        if gamma == 90.0:
+            cos_gamma = 0.0
+            sin_gamma = 1.0
+        else:
+            gamma = np.deg2rad(gamma)
+            cos_gamma = np.cos(gamma)
+            sin_gamma = np.sin(gamma)
+        box_matrix[1, 0] = ly * cos_gamma
+        box_matrix[1, 1] = ly * sin_gamma
+        box_matrix[2, 0] = lz * cos_beta
+        box_matrix[2, 1] = lz * (cos_alpha - cos_beta * cos_gamma) / sin_gamma
+        box_matrix[2, 2] = np.sqrt(lz * lz - box_matrix[2, 0] ** 2 - \
+                                   box_matrix[2, 1] ** 2)
+        # The discriminant of the above square root is only negative or zero for
+        # triplets of box angles that lead to an invalid box (i.e., the sum of
+        # any two angles is less than or equal to the third).
+        # We don't need to explicitly test for np.nan here since checking for a
+        # positive value already covers that.
+        if box_matrix[2, 2] > 0.0:
+            # all good, convert to correct dtype:
+            box_matrix = box_matrix.astype(np.float32)
+        else:
+            # invalid box, return zero vectors:
+            box_matrix = np.zeros((3, 3), dtype=np.float32)
+    return box_matrix
 
 
 def box_volume(dimensions):
-    """Return the volume of the unitcell described by *dimensions*.
+    """Return the volume of the unitcell described by `dimensions`.
 
-    The volume is computed as `det(x1,x2,x2)` where the xi are the
-    triclinic box vectors from :func:`triclinic_vectors`.
+    The volume is computed as *det(x1, x2, x3)* where the *xi* are the
+    triclinic box vectors obtained from :func:`triclinic_vectors`.
+    If the box is invalid, i.e., any box length is zero or negative, or any
+    angle is outside the open interval ``(0, 180)``, the resulting volume will
+    be zero.
 
     Parameters
     ----------
-    dimensions : [A, B, C, alpha, beta, gamma]
-        list of box lengths and angles (in degrees) such as
-        [A,B,C,alpha,beta,gamma]
+    dimensions : array_like
+        Unitcell dimensions provided in the same format as returned by
+        :attr:`MDAnalysis.coordinates.base.Timestep.dimensions`:\n
+        ``[lx, ly, lz, alpha, beta, gamma]``.
 
     Returns
     -------
     volume : float
+
+
+    .. versionchanged:: 0.20.0
+        Calculations are performed in double precision and zero is returned
+        for invalid dimensions.
     """
-    return np.linalg.det(triclinic_vectors(dimensions))
+    dim = np.asarray(dimensions, dtype=np.float64)
+    lx, ly, lz, alpha, beta, gamma = dim
+    if alpha == beta == gamma == 90.0 and lx > 0 and ly > 0 and lz > 0:
+        # valid orthogonal box, volume is the product of edge lengths:
+        volume = lx * ly * lz
+    else:
+        # triclinic or invalid box, volume is determinant of box matrix
+        # (invalid boxes are handled by triclinic_vectors):
+        volume = sarrus_det(triclinic_vectors(dim))
```

Comment:
If the boxes are always convex, I wonder how this performance compares with the volume calculation availble from Qhull via ConvexHull in SciPy

Reply:
@tylerjereddy 
> If the boxes are always convex

How would you construct a non-convex 3d shape from 3 linearly independent vectors?
> I wonder how this performance compares with the volume calculation available from Qhull via ConvexHull in SciPy

`scipy.spatial.ConvexHull` requires points as input, so you won't get around computing the triclinic box matrix from dimensions. From this matrix you then have to construct the remaining four corner points of the box (plus origin) and put those in a numpy array... What I want to say is that the determinant will probably be already computed before `ConvexHull` is even called.

Reply:
It is just a thought -- it seems the convexity assumption is solid as I expected, but you're likely right that there are too many operations in between.

The time required to test the two approaches probably isn't worth it, but intuition isn't always right for complexity analysis. If there was something faster in Qhull that just needed to be exposed in Cython I could always look into it.

Reply:
I just noticed that computing the determinant here doesn't make sense since `triclinic_vectors()` is guaranteed to return a lower triangular matrix. Therefore, the volume is simply the product of the diagonal.
`numpy.diag(matrix).prod()` is not an option since that takes 3.4 µs.
`matrix[0, 0] * matrix[1, 1] * matrix[2, 2]` takes 384 ns, and scipy *cannot* beat that simply due to the time the function call alone will take.
Not to mentiaon that this is an unnecessary micro-optimization anyway.

The reason for the Sarrus determinant is a different one; In the end I want to be able to compute large batches of 3x3 matrix inverses in a fast way, e.g., for computing the angular velocity vector from the moment of inertia and the angular momentum per molecule. `numpy.linalg.inv()` used to be horribly slow for that.

Reply:
hmm, some of that may depend on the linear algebra backend being used (MKL / openblas?). Anyway, it sounds like you're thinking about this very carefully which is great. Maybe I can follow up with some `asv` benchmarks at some point so we can be sure there's no performance regression on your hard work.